packages/react-reconciler/src/ReactFiberHydrationContext.js

```diff
@@ -106,8 +106,30 @@ function insertNonHydratedInstance(
 ) {
   fiber.effectTag |= Placement;
   if (__DEV__) {
-    // TODO: Find out what `hydrationWarningHostInstanceIndex` should be, `fiber.index` is wrong.
-    const hydrationWarningHostInstanceIndex = fiber.index;
+    let hydrationWarningHostInstanceIndex = 0;
+    {
+      // Count rendered host nodes by traversing `returnFiber` subtree until `fiber` is found.
+      let node = returnFiber.child;
+      const nextNodeStack = [];
+      while (node && node !== fiber) {
+        if (node.tag === HostComponent || node.tag === HostText) {
+          ++hydrationWarningHostInstanceIndex;
+        }
+        // Depth-first traversal.
+        if (node.child) {
+          if (node.sibling) {
+            // Remember where to continue on this tree level, then go deeper.
+            nextNodeStack.push(node.sibling);
+          }
+          node = node.child;
+        } else if (node.sibling) {
+          node = node.sibling;
+        } else {
+          node = nextNodeStack.pop();
+        }
+      }
+    }
```

Comment:
Can you please describe this algorithm in words? What is it looking for in particular?

Specifically I wonder if you can avoid the need for `nextNodeStack`.

Reply:
OK, I think I see what you're doing.

You're starting the search at the return fiber, and then looking for this fiber, counting how many host children there appear to be between them.

I think you can still avoid `nextNodeStack` by using `.return` pointers instead where necessary. However, you'll need to remember to write to them every time you descend down. See for example `propagateContextChange`.

Reply:
![sketch-1533143022802 01](https://user-images.githubusercontent.com/498274/43537020-f90c3898-9572-11e8-8c85-d45af77ae1a7.png)

Yes, I saw the traversal variant with writing into the return pointers but I thought it would be unsafe to modify the structure owned and maintained by another piece of code. If return pointers are only used during one traversal and their previous state is not relied upon, I can rewrite to use them instead of the stack.

Said that, this traversal is not performance-critical as it happens only in development mode and only when hydration failed.

Reply:
>If return pointers are only used during one traversal and their previous state is not relied upon, I can rewrite to use them instead of the stack.

Yeah, we rely on this pretty much everywhere.

>Said that, this traversal is not performance-critical as it happens only in development mode and only when hydration failed.

Agree but would prefer consistency since that's how we traverse elsewhere.